### PR TITLE
Upgraded to spago next and lastest version of psa-utils for iterative optimization 

### PR DESCRIPTION
**Pull request by avi892nash**
Updating purescript-psa with https://github.com/natefaubion/purescript-psa-utils/pull/13

### CODE REVIEW

File: spago.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    - ordered-collections
+    - versions
+    - debug
+workspace:
+  packageSet:
+    registry: 50.13.0
+  extraPackages: 
+    psa-utils: 
+      git: https://github.com/natefaubion/purescript-psa-utils.git
+      ref: master
```
Comment (avi892nash) [inline spago.yaml:27]: do not have a version tag in github. So, added a branch notation

File: index.js [hunk 2]
```diff
@@ -1,6 +1,5 @@
 #!/usr/bin/env node
-var pkg = require('./package.json');
-var Main$foreign = require('./output/Main/foreign.js');
-var Main = require('./output/Main');
-Main$foreign.version = pkg.version;
+import pkg from './package.json' assert { type: 'json' };
+import * as Main from './dist/index.js';
+global.version = pkg.version;
```
Comment (natefaubion) [inline index.js:4]: I don't think that there's a need to set a global variable here. We could have `main :: String -> Effect Unit` that takes the version as an argument.

Alternatively, we could bake in the version number like I did in purs-tidy https://github.com/natefaubion/purescript-tidy/blob/33910b845a5b72952c43084890b651fa657bd1ce/package.json#L17-L18 then we wouldn't need to load the package.json at all.

File: src/Main.purs [hunk 3]
```diff
@@ -179,7 +186,7 @@ main = void do
             else Stderr
     spawn' outputStream purs args \pursResult -> do
       for_ (Str.split (Str.Pattern "\n") pursResult.output) \err ->
-        case jsonParser err >>= decodeJson >>= parsePsaResult of
+        case jsonParser err >>= (either (Left <<< printJsonDecodeError) (Right <<< identity) <<< decodeJson) >>= parsePsaResult of
```
Comment (natefaubion) [inline src/Main.purs:189]: This could use `lmap printJsonDecodeError` here and later on.

File: src/Main.purs [hunk 4]
```diff
@@ -23,15 +25,20 @@ import Effect.Exception (catchException, throw, throwException)
...
 import Node.FS.Sync as File
 import Node.Path as Path
 import Node.Platform (Platform(Win32))
+import Node.Process (stderr, stdout)
 import Node.Process as Process
-import Node.Stream as Stream
-import Partial.Unsafe (unsafePartial)
+import Node.Stream (dataH)
+import Node.UnsafeChildProcess.Unsafe (spawn', unsafeStderr, unsafeStdout) as UnsafeChild
+import Node.UnsafeChildProcess.Safe (errorH, exitH) as UnsafeChild
```
Comment (natefaubion) [inline src/Main.purs:41]: Can you clarify what UnsafeChildProcess is about and why it's needed? I'm not up to date on everything that happened with childprocess.

File: src/Main.purs [hunk 5]
```diff
@@ -202,28 +209,29 @@ main = void do
...
     buffer <- Ref.new ""
-    fillBuffer buffer $
-      case outputStream of
-        Stdout -> Child.stdout child
-        Stderr -> Child.stderr child
-    Child.onExit child \status ->
+    _ <- maybe (pure $ pure unit) (fillBuffer buffer) (toMaybe $
+          case outputStream of
+            Stdout -> UnsafeChild.unsafeStdout child
+            Stderr -> UnsafeChild.unsafeStderr child)
```
Comment (natefaubion) [inline src/Main.purs:220]:
```suggestion
    traverse_ (fillBuffer buffer) $ toMaybe case outputStream of
      Stdout -> UnsafeChild.unsafeStdout child
      Stderr -> UnsafeChild.unsafeStderr child)
```

File: src/Main.purs [hunk 6]
```diff
@@ -272,7 +280,7 @@ main = void do
         let source = Array.slice (pos.startLine - 1) (pos.endLine) contents
         pure $ Just source
 
-  decodeStash s = jsonParser s >>= decodeJson >>= traverse parsePsaError
+  decodeStash s = jsonParser s >>= (either (Left <<< printJsonDecodeError) (Right <<< identity) <<< decodeJson) >>= (traverse parsePsaError)
```
Comment (natefaubion) [inline src/Main.purs:283]: This can be cleaned up too.

**Review by natefaubion — CHANGES_REQUESTED** [review]

File: src/Main.purs (hunk 4 above)
Comment (avi892nash) [reply src/Main.purs]: In the latest Node.ChildProcess, appendStdio is there in spawn', which appends stdio to [pipe, pipe, pipe], causing issues with seeing the "Compiling Module" message.

So, I'm using Node.UnsafeChildProcess.Unsafe.spawn' instead:

```
let stdio =
      case outputStream of
        Stdout -> [pipe, pipe, shareStream stderr]
        Stderr -> [pipe, shareStream stdout, pipe]
```
This way, the streams are configured correctly. Here's the [PursuitLink](https://pursuit.purescript.org/packages/purescript-node-child-process/11.1.0/docs/Node.ChildProcess).

File: spago.yaml (hunk 1 above)
Comment (natefaubion) [reply spago.yaml:27]: I added a v8.1.0 tag.

File: package.json [hunk 7]
```diff
@@ -9,23 +10,31 @@
...
+    "compile-self": "node index.js -c \"src/**/*.purs\" \".spago/p/**/src/**/*.purs\"",
+    "pretest": "rm -rf testcase/output",
+    "test": "node index.js --is-lib=testcase/bower_components --is-lib=testcase/lib -o \"testcase/output\" \"testcase/*.purs\" \"testcase/lib/*.purs\" \"testcase/bower_components/purescript-*/src/**/*.purs\"",
+    "pretest-purs": "rm -rf testcase/output",
+    "test-purs": "purs compile --output=testcase/output \"testcase/*.purs\" \"testcase/bower_components/purescript-*/src/**/*.purs\"",
+    "prepare": "rm -rf output && npm run compile"
   },
   "devDependencies": {
-    "bower": "^1.7.2"
+    "bower": "^1.8.14",
```
Comment (natefaubion) [inline package.json:36]: I think we need an esbuild dependency for spago bundle to work.

File: src/Main.purs [hunk 8]
```diff
@@ -202,28 +210,28 @@ main = void do
   spawn' outputStream cmd args onExit = do
     let stdio =
           case outputStream of
-            Stdout -> [ Just Child.Pipe, Just Child.Pipe, unsafePartial (Array.unsafeIndex Child.inherit 2) ]
-            Stderr -> [ Just Child.Pipe, unsafePartial (Array.unsafeIndex Child.inherit 1), Just Child.Pipe ]
+            Stdout -> [ pipe, pipe, shareStream stderr ]
+            Stderr -> [ pipe, shareStream stdout, pipe ]
 
-    child <- Child.spawn cmd args Child.defaultSpawnOptions { stdio = stdio }
+    child <- UnsafeChild.spawn' cmd args { stdio, detached : false }
```
Comment (natefaubion) [inline src/Main.purs:216]:
```suggestion
    child <- UnsafeChild.spawn' cmd args { stdio, detached: false }
```

File: src/Main.purs [hunk 9]
```diff
@@ -272,7 +280,7 @@ main = void do
         let source = Array.slice (pos.startLine - 1) (pos.endLine) contents
         pure $ Just source
 
-  decodeStash s = jsonParser s >>= decodeJson >>= traverse parsePsaError
+  decodeStash s = jsonParser s >>= (lmap printJsonDecodeError <<< decodeJson) >>= (traverse parsePsaError)
```
Comment (natefaubion) [inline src/Main.purs:283]:
```suggestion
  decodeStash s = jsonParser s >>= (lmap printJsonDecodeError <<< decodeJson) >>= traverse parsePsaError
```

File: package.json (hunk 7 above)
Comment (avi892nash) [reply package.json:36]: Done

**Comment by natefaubion** [comment]
Thanks for doing this!

**Pull request merged**